### ISSUE TITLE
Function to get number of files / available files for OSFileStore

### ISSUE THREAD

**Issue by Space2Man** [oshi/oshi#617]
Hi,

I would like to use OSHI to monitor the utilisation of filesystems. Regarding disk space utilisation this is easily possible with the metrics provided by OSFileStore. But on Unix systems it's also important to monitor the number of existing files/inodes vs. the number of total files/inodes allowed in the filesystem.

Is it possible to include such a metric in OSHI even though Windows does not have this notion? I would not mind if it would return -1 on Windows for both metrics.

Thanks a lot for this great library! It's really useful!

Space

**Comment by retomerz** [comment]
> Is it possible to include such a metric in OSHI even though Windows does not have this notion? I would not mind if it would return -1 on Windows for both metrics.

Instead of return wrong (0) or invalid values (-1) the implementation should throw a java.lang.UnsupportedOperationException
And the API should have a "isSupported(Feature)" methode.
This is also how a modern API in JRE handles it. For example see
https://docs.oracle.com/javase/9/docs/api/java/awt/Taskbar.html#isSupported-java.awt.Taskbar.Feature-

**Comment by cilki** [comment]
I'm working on a design pattern that would allow platform specific features to be added cleanly without exceptions or invalid return values. It may not be ready for a while, but we will definitely include this along with many other specific features eventually.

**Comment by dbwiddis** [comment]
@Space2Man What is the difference between what you're proposing and the existing [`FileSystem.getOpenFileDescriptors()`](http://oshi.github.io/oshi/apidocs/oshi/software/os/FileSystem.html#getOpenFileDescriptors--) and [`FileSystem.getMaxFileDescriptors()`](http://oshi.github.io/oshi/apidocs/oshi/software/os/FileSystem.html#getMaxFileDescriptors--) methods?  (Which, by the way, Windows returns open/max Handles which is a related but totally different metric but better than -1.) 

@retomerz As @cilki points out we definitely plan on improving the way we handle this in the next major version, considering UnsupportedExceptions, or simply `null` results, or possibly [Optional](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) objects all being considered.  This is why I need SDEs on the project.  Do please join in the [design discussions](https://github.com/oshi/oshi/issues?q=is%3Aissue+is%3Aopen+label%3A%22design+discussion%22) and help shape the future of OSHI.

**Comment by Space2Man** [comment]
@dbwiddis My request is about the number of files that are stored in a filesystem and the maximum number of files that can be stored in a filesystem. On Windows this is not important but on Linux each file is handled as an inode and the number of inodes per FS is limited by the FS.

The informations can be checked by running df -i

excerpt from man-page:

```
      -i, --inodes
              list inode information instead of block usage
```

For a filesystem it might look like this:

```
# df -i /boot
Filesystem     Inodes IUsed  IFree IUse% Mounted on
/dev/md127     131072   343 130729    1% /boot
```

Thanks and best regards, Space

**Comment by dbwiddis** [comment]
Got it... now I understand the difference.  We prefer native functions rather than command-line parsing when such is available.  It looks like the information from `df -i` is available in [`statvfs()`](http://man7.org/linux/man-pages/man3/statvfs.3.html).  Can you put together a PR?

**Comment by dbwiddis** [comment]
Ok, I have to share.  The `statvfs()` man page I linked above worked for Linux. Here's what [`man statvfs`](https://www.freebsd.org/cgi/man.cgi?query=statvfs&manpath=FreeBSD+8.1-RELEASE) says on FreeBSD:

> The statvfs() and fstatvfs() functions fill the structure pointed to by buf with garbage. This garbage will occasionally bear resemblance to file system statistics, but portable applications must not depend on this.

**Comment by cilki** [comment]
> The statvfs() and fstatvfs() functions fill the structure pointed to by buf with garbage. This garbage will occasionally bear resemblance to file system statistics, but portable applications must not depend on this.

This sort of thing is why I never really liked systems programming 🤣

**Comment by cilki** [comment]
Maybe we should mark an API like that with an annotation so the issue tracker isn't flooded with people who unfortunately ran that command on a network share.

Something like `@SuperSlow`, `@ComputeIntensive`, etc...

**Comment by dbwiddis** [comment]
Interesting idea, although in this particular case I wouldn't waste the time writing slow code to give an answer that was only remotely related.  Also should consider multithreading slow requests and reading the result at some later point in time.  All good topics for another design thread issue!  For this one and OSHI 3.x, I think -1 is ok.

**Comment by Space2Man** [comment]
Do you have an example of calling such a native function?

**Comment by dbwiddis** [comment]
Look at any of the interfaces in the `oshi.jna.platform.*` packages.

Here's how OSHI did `statfs()` on Mac OS X:
Structure [defined here](https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/jna/platform/mac/SystemB.java#L254).    Array of those structures filled with [this native call](https://github.com/oshi/oshi/blob/a73f04a1f2eec9eb5e7a59cebd1093f879d1cec0/oshi-core/src/main/java/oshi/jna/platform/mac/SystemB.java#L565).  Called from code [like this](https://github.com/oshi/oshi/blob/a73f04a1f2eec9eb5e7a59cebd1093f879d1cec0/oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java#L91).

You would do something similar for `statvfs()`... define the structure in an interface that extends JNA's `Library` class.  You'd put it in `oshi.jna.platform.linux.Libc.java`.  You'd map the parameters from the structure on the [man page](http://man7.org/linux/man-pages/man3/statvfs.3.html), (use `NativeLong` for all the types since they're `unsigned long` and could be 32 or 64 bit).  Then you'd define the native function as `int statvfs(String path, Statvfs buf);` (using the structure type for Statvfs).   Then you'd call it in code with 
```
Statvfs statvfs = new Statvfs(); 
Libc.INSTANCE.statvfs("\path\here", statvfs);
```
At which point you should be able to access `statvfs.f_files` etc.

**Comment by Space2Man** [comment]
Thanks, I had found the calls in Libc.java after I had sent the comment ... I will see what i can do.

**Comment by Space2Man** [comment]
Not sure how to set up the structure because of this _f_unused which is only added if word size is 32bit ... 

```
#if (__WORDSIZE == 32 \
     && (!defined __SYSCALL_WORDSIZE || __SYSCALL_WORDSIZE == 32))
#define _STATVFSBUF_F_UNUSED
#endif

struct statvfs
  {
    unsigned long int f_bsize;
    unsigned long int f_frsize;
#ifndef __USE_FILE_OFFSET64
    __fsblkcnt_t f_blocks;
    __fsblkcnt_t f_bfree;
    __fsblkcnt_t f_bavail;
    __fsfilcnt_t f_files;
    __fsfilcnt_t f_ffree;
    __fsfilcnt_t f_favail;
#else
    __fsblkcnt64_t f_blocks;
    __fsblkcnt64_t f_bfree;
    __fsblkcnt64_t f_bavail;
    __fsfilcnt64_t f_files;
    __fsfilcnt64_t f_ffree;
    __fsfilcnt64_t f_favail;
#endif
    unsigned long int f_fsid;
#ifdef _STATVFSBUF_F_UNUSED
    int __f_unused;
#endif
    unsigned long int f_flag;
    unsigned long int f_namemax;
    int __f_spare[6];
  };
```

Also in a commit for JNA I saw a response from you that getFieldOrder() is deprecated. 

Access to the infos from statvfs is working on *my* system. But not sure if that is working on e.g. a 32bit Linux system. Also I have no access to FreeBSD or Solaris or MacOS ...

**Comment by Space2Man** [comment]
Would you be ok with something like this for the structure?
```
    class Statvfs extends Structure {
        private static final int PADDING_SIZE = 8 / NativeLong.SIZE - 1;

        public NativeLong fsBlockSize;
        public NativeLong fsFragmentSize;
        public NativeLong fsSizeInBlocks;
        public NativeLong fsBlocksFree;
        public NativeLong fsBlocksFreeUnpriv;
        public NativeLong fsTotalInodeCount;
        public NativeLong fsFreeInodeCount;
        public NativeLong fsFreeInodeCountUnpriv;
        public NativeLong fsId;
        public int[] _funused = new int[PADDING_SIZE];  // Won't be written for 64-bit systems
        public NativeLong fsMountFlags;
        public NativeLong fsMaxFilenameLength;
        public int[] _fspare = new int[6];

        @Override
        protected List<Field> getFieldList() {
            System.out.println("FieldList: PADDING_SIZE: " + PADDING_SIZE);
            List<Field> fields = new ArrayList<Field>(super.getFieldList());
            if (PADDING_SIZE == 0) {
                Iterator<Field> fieldIterator = fields.iterator();
                while (fieldIterator.hasNext()) {
                    Field field = fieldIterator.next();
                    if ("_funused".equals(field.getName())) {
                        fieldIterator.remove();
                    }
                }
            }
            return fields;
        }

        protected List<String> getFieldOrder() {
            if (PADDING_SIZE == 0) {
                return Arrays.asList(new String[]{"fsBlockSize", "fsFragmentSize", "fsSizeInBlocks", "fsBlocksFree",
                        "fsBlocksFreeUnpriv", "fsTotalInodeCount", "fsFreeInodeCount", "fsFreeInodeCountUnpriv",
                        "fsId", "fsMountFlags", "fsMaxFilenameLength", "_fspare"});
            } else {
                return Arrays.asList(new String[]{"fsBlockSize", "fsFragmentSize", "fsSizeInBlocks", "fsBlocksFree",
                        "fsBlocksFreeUnpriv", "fsTotalInodeCount", "fsFreeInodeCount", "fsFreeInodeCountUnpriv",
                        "fsId", "_funused", "fsMountFlags", "fsMaxFilenameLength", "_fspare"});
            }
        }

    }

    int statvfs(String path, Statvfs buf);
```

**Comment by dbwiddis** [comment]
Yes.  Sorry I was too busy to get back to you. The way we eventually handled that JNA commit you referenced is probably the best.  (And you should submit the same change to JNA!)

**Comment by dbwiddis** [comment]
Also I have a 32 bit Linux VM to test with.  You can write the code on AbstractOoeratingSystem and test it on Linux. I can test it on the other systems when you commit it.

**Comment by Space2Man** [comment]
Hi, when trying to test I get the following message:

`[ERROR] Failed to execute goal com.mycila:license-maven-plugin:3.0:format (default) on project oshi-core: Execution default of goal com.mycila:license-maven-plugin:3.0:format failed: Cannot read header document LICENSE_HEADER. Cause: Resource LICENSE_HEADER not found in file system, classpath or URL: no protocol: LICENSE_HEADER -> [Help 1]`

Do you get that as well?

**Comment by dbwiddis** [comment]
I was going to suggest update/clean.  One of my first steps always. I don't know what happens but it's kinda like rebooting your machine. :)

Also FYI I bumped OSHI up to JNA 5.0.0.  This introduces the @FieldOrder annotation which will simplify your structure.  You can just define the longer version and remove the `_funused` in the zero padding case in the two overrides.

**Comment by dbwiddis** [comment]
Merged in #659
